### PR TITLE
Record name can now be specified to the AVRO `_infer_schema` function

### PR DESCRIPTION
**Pull request by tristan0x**
### Use case

AVRO files written by this library are used in a Java Spark job:
* org.apache.avro:avro:1.8.1
* com.databricks:spark-avro_2.10:2.0.1
* org.apache.spark:spark-core_2.10:1.6.0

I got exceptions because the schema does not have a record name.

### What does it do?

It augment capabilities of `hdfs.ext.avro._infer_schema` function so that it is able to provide a name to the encountered records.

### Design
As `AvroWriter` has a `schema` parameter, I could compute the schema by myself and provide it to the `AvroWriter` but I believe it is cleaner to infer it when the first element is sent to `AvroWriter._write`.

If you don't like it, would you consider an additional `infer_schema_func` parameter to the `AvriWriter` constructor ?

### CODE REVIEW

**Comment by mtth** [comment]
Thanks for the new pull request!

The `record_name` option is probably too specific to be exposed on the `AvroWriter`. Do you care about being able to choose the name or only about having one to be able to read the data from Spark?

File: hdfs/ext/avro/__init__.py [hunk 1]
```diff
@@ -64,18 +65,22 @@ def _infer_schema(obj):
...
     return {
       'type': 'array',
-      'items': _infer_schema(obj[0])
+      'items': _infer_schema(obj[0], record_name=record_name)
     }
   elif isinstance(obj, dict):
     if not obj:
       raise ValueError('Cannot infer type of empty record.')
-    return {
+    nested_record_name = u'Nested' + record_name if record_name else None
```
Comment (mtth) [inline hdfs/ext/avro/__init__.py]: This could cause duplicate names if there are multiple records at the same nesting level.

Comment (tristan0x) [inline hdfs/ext/avro/__init__.py]: Yeah I assumed that records have the same type at a given level. I did not have the chance to test it so I do not know how Avro reacts in this case.

**Comment by tristan0x** [comment]
Thanks for the quick feedback @mtth 
Indeed, I don't really mind what the record type name is as long as there is one.

File: hdfs/ext/avro/__init__.py (hunk 1 above)
Comment (mtth) [inline hdfs/ext/avro/__init__.py]: Avro would complain in this case (names should be unique).

**Comment by mtth** [comment]
Makes sense. The best option is probably to generate a random name for each record then. It'll keep the `AvroWriter` API simple and avoid duplicate name problems. How about a pattern like `__Record1` where the suffix gets incremented each time?

**Comment by mtth** [comment]
Implemented and released in `2.0.10`.

**Pull request closed without merging**